File: backend/routers/debug.py
"""
Debug and monitoring endpoints.

These endpoints provide access to cache statistics and other debugging information.
"""

from typing import Dict

from fastapi import APIRouter
from services.cache_service import get_cache_service

router = APIRouter()


@router.get("/cache/stats")
async def get_cache_stats() -> Dict[str, int]:
    """
    Get current cache statistics.

    Returns:
        Dictionary containing:
        - hits: Number of cache hits
        - misses: Number of cache misses
        - total_requests: Total cache requests
        - hit_rate: Cache hit rate percentage
        - size: Number of entries in cache
        - invalidations: Number of cache invalidations
    """
    cache_service = get_cache_service()
    return cache_service.get_stats()


@router.post("/cache/cleanup")
async def cleanup_cache() -> Dict[str, str]:
    """
    Manually trigger cache cleanup to remove expired entries.

    Returns:
        Success message
    """
    cache_service = get_cache_service()
    cache_service.cleanup_expired()
    return {"status": "success", "message": "Cache cleanup completed"}


@router.post("/cache/clear")
async def clear_cache() -> Dict[str, str]:
    """
    Clear all cache entries.

    WARNING: This will clear the entire cache and may temporarily impact performance.

    Returns:
        Success message
    """
    cache_service = get_cache_service()
    cache_service.clear()
    return {"status": "success", "message": "Cache cleared successfully"}

File: backend-ts/src/http/routes/debug.ts
/**
 * Cache statistics and manual cache maintenance.
 *
 * Ported from `backend/routers/debug.py`, mounted under `/debug`. Python routes
 * these through `services/cache_service.py`, which for all three endpoints is a
 * pass-through to the cache manager — so they go straight to `getCache()` here
 * rather than growing a service layer whose only job would be to forward.
 *
 * Nothing in `frontend/` calls these; they exist for poking at a running
 * backend. They are still behind auth, exactly as in Python — `/debug` is on
 * neither exclusion list.
 */

import { Hono } from 'hono'

import { getCache } from '../../infrastructure/cache'
import { getLogger } from '../../infrastructure/logging/logger'
import type { AppEnv } from '../types'

const logger = getLogger('Debug')

/**
 * The wire shape of `CacheManager.get_stats()`.
 *
 * `src/infrastructure/cache.ts` keeps its counters in camelCase because it is
 * an internal module; Python's dict keys are what actually go out over HTTP, so
 * the rename happens here and only here.
 */
interface CacheStatsResponse {
  hits: number
  misses: number
  invalidations: number
  evictions: number
  total_requests: number
  hit_rate: number
  size: number
}

export function createDebugRoutes(): Hono<AppEnv> {
  const routes = new Hono<AppEnv>()

  routes.get('/debug/cache/stats', (c) => {
    const stats = getCache().getStats()
    const body: CacheStatsResponse = {
      hits: stats.hits,
      misses: stats.misses,
      invalidations: stats.invalidations,
      evictions: stats.evictions,
      total_requests: stats.totalRequests,
      hit_rate: stats.hitRate,
      size: stats.size,
    }
    return c.json(body)
  })

  routes.post('/debug/cache/cleanup', (c) => {
    getCache().cleanupExpired()
    logger.info('Cache cleanup completed')
    return c.json({ status: 'success', message: 'Cache cleanup completed' })
  })

  routes.post('/debug/cache/clear', (c) => {
    getCache().clear()
    logger.info('Cache cleared')
    return c.json({ status: 'success', message: 'Cache cleared successfully' })
  })

  return routes
}
